pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java

```diff
@@ -425,28 +426,31 @@ private void addMatchingChildrenToQueue(Queue<SearchEntry> searchQueue, StarTree
 
       int[] matchingDictionaryIds = predicateEntry.predicateEvaluator.getMatchingDictionaryIds();
       for (int matchingDictionaryId : matchingDictionaryIds) {
-        if (children.containsKey(matchingDictionaryId)) {
-          addNodeToSearchQueue(searchQueue, children.get(matchingDictionaryId), remainingPredicateColumns,
-              remainingGroupByColumns);
+        StarTreeIndexNodeInterf child = node.getChildForDimensionValue(matchingDictionaryId);
+        if (child != null) {
+          addNodeToSearchQueue(searchQueue, child, remainingPredicateColumns, remainingGroupByColumns);
         }
       }
     } else {
       int nextValueId;
       if (groupByColumns.contains(column) || predicatesMap.containsKey(column)
-          || !children.containsKey(StarTreeIndexNode.all())) {
-        for (StarTreeIndexNode indexNode : children.values()) {
-          if (indexNode.getDimensionValue() != StarTreeIndexNode.all()) {
+          || (node.getChildForDimensionValue(StarTreeIndexNodeInterf.ALL) == null)) {
+        Iterator<? extends StarTreeIndexNodeInterf> childrenIterator = node.getChildrenIterator();
```

Comment:
Because iterator of super class (or interface) is not type compatible with iterator of sub class (implementation class).